File: colis-direct/ios-native/ColisDirectApp/Sources/Data/API/APIService.swift
import Foundation

// MARK: - API Error
enum APIError: Error, LocalizedError {
    case invalidURL
    case networkError(Error)
    case serverError(Int, String?)
    case decodingError(Error)
    case unauthorized
    case unknown

    var errorDescription: String? {
        switch self {
        case .invalidURL: return "URL invalide"
        case .networkError(let e): return "Erreur réseau: \(e.localizedDescription)"
        case .serverError(_, let msg): return msg ?? "Erreur serveur"
        case .decodingError: return "Erreur de décodage"
        case .unauthorized: return "Session expirée. Veuillez vous reconnecter."
        case .unknown: return "Erreur inconnue"
        }
    }
}

// MARK: - API Service
final class APIService: @unchecked Sendable {
    nonisolated(unsafe) static let shared = APIService()
    private init() {}

    private let tokenManager = TokenManager.shared
    private let baseURL = AppConfig.baseURL

    private var session: URLSession {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        return URLSession(configuration: config)
    }

    // MARK: - Generic Request
    private func request<T: Decodable>(
        endpoint: String,
        method: String = "GET",
        body: Encodable? = nil,
        authenticated: Bool = true
    ) async throws -> T {
        guard let url = URL(string: baseURL + endpoint) else {
            throw APIError.invalidURL
        }

        var req = URLRequest(url: url)
        req.httpMethod = method
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")

        if authenticated, let token = tokenManager.getToken() {
            req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        if let body = body {
            req.httpBody = try JSONEncoder().encode(body)
        }

        let (data, response) = try await session.data(for: req)

        guard let http = response as? HTTPURLResponse else {
            throw APIError.unknown
        }

        if http.statusCode == 401 {
            if req.url?.path.contains("auth/signin") == true {
                let msg = try? JSONDecoder().decode(ErrorResponse.self, from: data)
                let errorText = msg?.error ?? msg?.message ?? "Identifiants incorrects"
                throw APIError.serverError(http.statusCode, errorText)
            }
            tokenManager.clearAll()
            throw APIError.unauthorized
        }

        if !(200..<300).contains(http.statusCode) {
            let msg = try? JSONDecoder().decode(ErrorResponse.self, from: data)
            let errorText = msg?.error ?? msg?.message
            throw APIError.serverError(http.statusCode, errorText)
        }

        do {
            let decoder = JSONDecoder()
            return try decoder.decode(T.self, from: data)
        } catch {
            throw APIError.decodingError(error)
        }
    }

    // MARK: - Auth
    func signIn(email: String? = nil, phone: String? = nil, password: String) async throws -> AuthResponse {
        let body = SignInRequest(email: email, phone: phone, password: password)
        let response: AuthResponse = try await request(endpoint: "auth/signin", method: "POST", body: body, authenticated: false)
        tokenManager.saveToken(response.accessToken)
        tokenManager.saveRole(response.user.role)
        return response
    }

    func signUp(email: String, password: String, firstName: String, lastName: String, phone: String) async throws -> AuthResponse {
        let body = SignUpRequest(email: email, password: password, firstName: firstName, lastName: lastName, phone: phone)
        let response: AuthResponse = try await request(endpoint: "auth/signup", method: "POST", body: body, authenticated: false)
        tokenManager.saveToken(response.accessToken)
        tokenManager.saveRole(response.user.role)
        return response
    }

    func getMe() async throws -> UserDto {
        try await request(endpoint: "auth/me")
    }

    func signOut() async throws {
        let _: SuccessResponse = try await request(endpoint: "auth/signout", method: "POST")
        tokenManager.clearAll()
    }

    // MARK: - Shipments
    func getShipments(status: String? = nil, paymentStatus: String? = nil) async throws -> [ShipmentDto] {
        var endpoint = "shipments"
        var params: [String] = []
        if let s = status { params.append("current_status=\(s)") }
        if let p = paymentStatus { params.append("payment_status=\(p)") }
        if !params.isEmpty { endpoint += "?" + params.joined(separator: "&") }
        return try await request(endpoint: endpoint)
    }

    func getShipmentById(_ id: String) async throws -> ShipmentDto {
        try await request(endpoint: "shipments/\(id)")
    }

    func createShipment(_ req: CreateShipmentRequest) async throws -> ShipmentDto {
        try await request(endpoint: "shipments", method: "POST", body: req)
    }

    func cancelShipment(trackingNumber: String) async throws -> SuccessResponse {
        try await request(endpoint: "shipments/\(trackingNumber)/cancel", method: "POST")
    }

    func switchToRelayPayment(trackingNumber: String) async throws -> SuccessResponse {
        try await request(endpoint: "shipments/\(trackingNumber)/switch-to-relay-payment", method: "POST")
    }

    func initPaystack(
        trackingNumber: String,
        amountFcfa: Double,
        customerName: String,
        customerEmail: String,
        customerPhone: String
    ) async throws -> PaystackInitResponse {
        let body = PaystackInitRequest(
            trackingNumber: trackingNumber,
            amountFcfa: amountFcfa,
            customerName: customerName,
            customerEmail: customerEmail,
            customerPhone: customerPhone
        )
        return try await request(endpoint: "payments/paystack/init", method: "POST", body: body)
    }

    func verifyPaystack(reference: String, trackingNumber: String) async throws -> PaystackVerifyResponse {
        let body = PaystackVerifyRequest(reference: reference, trackingNumber: trackingNumber)
        return try await request(endpoint: "payments/paystack/verify", method: "POST", body: body)
    }

    // MARK: - Tracking
    func getPublicTracking(trackingNumber: String) async throws -> TrackingResponse {
        try await request(endpoint: "tracking/\(trackingNumber)", authenticated: false)
    }

    func getTracking(trackingNumber: String) async throws -> TrackingResponse {
        try await request(endpoint: "tracking/\(trackingNumber)")
    }

    // MARK: - Relay Points
    func getRelayPoints() async throws -> [RelayPointDto] {
        try await request(endpoint: "relay-points", authenticated: false)
    }

    func getRelayPoint(_ id: String) async throws -> RelayPointDto {
        try await request(endpoint: "relay-points/\(id)")
    }

    // MARK: - Pricing
    func calculatePricing(senderCommune: String, recipientCommune: String, packageSize: String, weight: Double) async throws -> PricingCalculateResponse {
        let endpoint = "pricing-grids/calculate?sender_commune=\(senderCommune.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? senderCommune)&recipient_commune=\(recipientCommune.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? recipientCommune)&package_size=\(packageSize)&grid_type=colis&weight=\(weight)"
        return try await request(endpoint: endpoint, authenticated: false)
    }

    // MARK: - Scan (Relay)
    func relayIntake(trackingNumber: String) async throws -> ScanResponse {
        let body = RelayIntakeRequest(trackingNumber: trackingNumber)
        return try await request(endpoint: "scan/relay-intake", method: "POST", body: body)
    }

    func relayFinalIntake(trackingNumber: String) async throws -> ScanResponse {
        let body = RelayIntakeRequest(trackingNumber: trackingNumber)
        return try await request(endpoint: "scan/relay-final-intake", method: "POST", body: body)
    }

    func makeAvailable(trackingNumber: String) async throws -> ScanResponse {
        let body = RelayIntakeRequest(trackingNumber: trackingNumber)
        return try await request(endpoint: "scan/ops/make-available", method: "POST", body: body)
    }

    func completeDelivery(trackingNumber: String, recipientPhone: String) async throws -> ScanResponse {
        let body = CompleteDeliveryRequest(trackingNumber: trackingNumber, recipientPhone: recipientPhone)
        return try await request(endpoint: "scan/relay/complete-delivery", method: "POST", body: body)
    }

    // MARK: - Scan (Transporter)
    func carrierPickup(trackingNumber: String) async throws -> ScanResponse {
        let body = CarrierPickupRequest(trackingNumber: trackingNumber)
        return try await request(endpoint: "scan/carrier-pickup", method: "POST", body: body)
    }

    func confirmHomePickup(trackingNumber: String, senderPhone: String) async throws -> ScanResponse {
        let body = HomePickupConfirmRequest(trackingNumber: trackingNumber, senderPhone: senderPhone)
        return try await request(endpoint: "scan/confirm-home-pickup", method: "POST", body: body)
    }

    // MARK: - Transporter Handoffs
    func getTransporterAssignments() async throws -> [ShipmentDto] {
        try await request(endpoint: "handoffs/transporter/assignments")
    }

    func getDeliveredShipments() async throws -> [ShipmentDto] {
        try await request(endpoint: "handoffs/transporter/delivered-shipments")
    }

    // MARK: - Search
    func searchByPhone(_ phone: String) async throws -> [ShipmentDto] {
        try await request(endpoint: "shipments/search/phone/\(phone)")
    }

    // MARK: - Shipping Addresses (expéditeur)
    func getShippingAddresses() async throws -> [ShippingAddressDto] {
        try await request(endpoint: "shipping-addresses")
    }

    // MARK: - Recipient Addresses (carnet d'adresses)
    func getRecipientAddresses() async throws -> [RecipientAddressDto] {
        try await request(endpoint: "recipient-addresses")
    }
}

// MARK: - Error Response
private struct ErrorResponse: Codable {
    let error: String?
    let message: String?
}

File: colis-direct/android/app/src/main/java/ci/colisdirect/app/data/api/ApiService.kt
package ci.colisdirect.app.data.api

import ci.colisdirect.app.data.api.model.*
import retrofit2.Response
import retrofit2.http.*

interface ApiService {

    // ===================== AUTH =====================

    @POST("auth/signin")
    suspend fun signIn(@Body request: SignInRequest): Response<AuthResponse>

    @POST("auth/signup")
    suspend fun signUp(@Body request: SignUpRequest): Response<AuthResponse>

    @GET("auth/me")
    suspend fun getMe(): Response<MeResponse>

    @PATCH("users/{id}")
    suspend fun updateUser(
        @Path("id") id: String,
        @Body request: UpdateUserRequest,
    ): Response<UserDto>

    @POST("auth/signout")
    suspend fun signOut(): Response<SuccessResponse>

    // ===================== SHIPMENTS =====================

    @GET("shipments")
    suspend fun getShipments(
        @Query("current_status") currentStatus: String? = null,
        @Query("payment_status") paymentStatus: String? = null,
        @Query("relay_id") relayId: String? = null,
    ): Response<List<ShipmentDto>>

    @GET("shipments/{id}")
    suspend fun getShipmentById(@Path("id") id: String): Response<ShipmentDto>

    @POST("shipments")
    suspend fun createShipment(@Body request: CreateShipmentRequest): Response<ShipmentDto>

    @POST("promo-codes/validate")
    suspend fun validatePromoCode(@Body request: PromoValidateRequest): Response<PromoValidateEnvelope>

    @GET("shipments/search/phone/{phone}")
    suspend fun searchByPhone(@Path("phone") phone: String): Response<List<ShipmentDto>>

    @GET("shipments/pickup/tracking/{trackingNumber}")
    suspend fun getShipmentForPickup(
        @Path("trackingNumber") trackingNumber: String
    ): Response<ShipmentDto>

    @GET("shipments/pickup/sender-phone/{phone}")
    suspend fun getShipmentsForHomePickup(
        @Path("phone") phone: String
    ): Response<List<ShipmentDto>>

    // ===================== SCAN =====================

    @POST("scan/relay-intake")
    suspend fun relayIntake(@Body request: RelayIntakeRequest): Response<ScanResponse>

    @POST("scan/carrier-pickup")
    suspend fun carrierPickup(@Body request: CarrierPickupRequest): Response<ScanResponse>

    @POST("scan/confirm-home-pickup")
    suspend fun confirmHomePickup(@Body request: HomePickupConfirmRequest): Response<ScanResponse>

    @POST("scan/relay-final-intake")
    suspend fun relayFinalIntake(@Body request: RelayIntakeRequest): Response<ScanResponse>

    @POST("scan/ops/make-available")
    suspend fun makeAvailable(@Body request: RelayIntakeRequest): Response<ScanResponse>

    @POST("scan/relay/complete-delivery")
    suspend fun completeDelivery(@Body request: CompleteDeliveryRequest): Response<ScanResponse>

    @POST("scan/ops/departure")
    suspend fun departure(@Body request: RelayIntakeRequest): Response<ScanResponse>

    // ===================== TRACKING =====================

    @GET("tracking/{trackingNumber}")
    suspend fun getPublicTracking(
        @Path("trackingNumber") trackingNumber: String
    ): Response<TrackingResponse>

    @GET("tracking/{trackingNumber}")
    suspend fun getTracking(
        @Path("trackingNumber") trackingNumber: String
    ): Response<TrackingResponse>

    // ===================== RELAY POINTS =====================

    @GET("relay-points")
    suspend fun getRelayPoints(): Response<List<RelayPointDto>>

    @GET("relay-points/{id}")
    suspend fun getRelayPoint(@Path("id") id: String): Response<RelayPointDto>

    @GET("relay-points/me")
    suspend fun getMyRelayPoint(): Response<RelayPointDto>

    @GET("relay-points/{id}/stats")
    suspend fun getRelayPointStats(@Path("id") id: String): Response<RelayStatsDto>

    @GET("relay-points/{id}/active-shipments")
    suspend fun getRelayActiveShipments(@Path("id") id: String): Response<List<ShipmentDto>>

    // ===================== RECIPIENT ADDRESSES =====================

    @GET("recipient-addresses")
    suspend fun getRecipientAddresses(): Response<List<RecipientAddressDto>>

    @POST("recipient-addresses")
    suspend fun createRecipientAddress(@Body request: CreateRecipientAddressRequest): Response<RecipientAddressDto>

    @DELETE("recipient-addresses/{id}")
    suspend fun deleteRecipientAddress(@Path("id") id: String): Response<SuccessResponse>

    @POST("transporter-applications")
    suspend fun submitTransporterApplication(
        @Body request: TransporterApplicationRequest,
    ): Response<PartnerApplicationResponse>

    @POST("relay-applications")
    suspend fun submitRelayApplication(
        @Body request: RelayApplicationRequest,
    ): Response<PartnerApplicationResponse>

    // ===================== PRICING =====================

    @GET("pricing-grids/calculate")
    suspend fun calculatePricing(
        @Query("sender_commune") senderCommune: String,
        @Query("recipient_commune") recipientCommune: String,
        @Query("package_size") packageSize: String,
        @Query("grid_type") gridType: String = "colis",
        @Query("weight") weight: Double,
    ): Response<PricingCalculateResponse>

    // ===================== HANDOFFS (Transporter) =====================

    @GET("handoffs/transporter/assignments")
    suspend fun getTransporterAssignments(): Response<List<ShipmentDto>>

    @GET("handoffs/transporter/delivered-shipments")
    suspend fun getDeliveredShipments(): Response<List<ShipmentDto>>

    @GET("handoffs/transporter/profile")
    suspend fun getTransporterProfile(): Response<TransporterProfileDto>

    @GET("transporter/wallet")
    suspend fun getTransporterWallet(): Response<TransporterWalletResponse>

    @GET("transporter/wallet/transactions")
    suspend fun getWalletTransactions(
        @Query("limit") limit: Int,
        @Query("offset") offset: Int,
    ): Response<WalletTransactionsResponse>

    @POST("transporter/wallet/withdraw")
    suspend fun requestWithdrawal(@Body request: WithdrawalRequest): Response<SimpleSuccessResponse>

    @GET("delivery-offers/my-offers")
    suspend fun getMyDeliveryOffers(): Response<List<DeliveryOfferDto>>

    @POST("delivery-offers/{offerId}/accept")
    suspend fun acceptDeliveryOffer(@Path("offerId") offerId: String): Response<SimpleSuccessResponse>

    @POST("delivery-offers/{offerId}/decline")
    suspend fun declineDeliveryOffer(@Path("offerId") offerId: String): Response<SimpleSuccessResponse>

    // ===================== PAYMENT =====================

    @POST("payments/paystack/init")
    suspend fun initPaystack(@Body request: PaystackInitRequest): Response<PaystackInitResponse>

    @POST("payments/paystack/verify")
    suspend fun verifyPaystack(@Body request: PaystackVerifyRequest): Response<PaystackVerifyResponse>

    @GET("payments/automated/{trackingNumber}")
    suspend fun getAutomatedPayment(
        @Path("trackingNumber") trackingNumber: String,
    ): Response<AutomatedPaymentDto>

    @GET("tracking/{trackingNumber}/dispatch-status")
    suspend fun getDispatchStatus(
        @Path("trackingNumber") trackingNumber: String,
    ): Response<DispatchStatusDto>

    @POST("payments/relay-cash/confirm")
    suspend fun confirmRelayCash(@Body request: RelayCashConfirmRequest): Response<SuccessResponse>

    @POST("shipments/{trackingNumber}/deliver")
    suspend fun deliverShipment(
        @Path("trackingNumber") trackingNumber: String,
        @Body request: DeliverShipmentRequest,
    ): Response<DeliverShipmentResponse>

    @POST("shipments/{trackingNumber}/cancel")
    suspend fun cancelShipment(@Path("trackingNumber") trackingNumber: String): Response<SuccessResponse>

    @POST("shipments/{trackingNumber}/switch-to-relay-payment")
    suspend fun switchToRelayPayment(@Path("trackingNumber") trackingNumber: String): Response<SuccessResponse>

    // ===================== ADMIN / SUPPORT =====================

    @GET("stats")
    suspend fun getAdminStats(): Response<AdminDashboardStatsDto>

    @GET("users")
    suspend fun getUsers(
        @Query("role") role: String? = null,
        @Query("search") search: String? = null,
    ): Response<List<UserDto>>

    @GET("relay-applications")
    suspend fun getRelayApplications(
        @Query("status") status: String? = null,
    ): Response<List<RelayApplicationDto>>

    @GET("transporter-applications")
    suspend fun getTransporterApplications(
        @Query("status") status: String? = null,
    ): Response<List<TransporterApplicationDto>>

    @GET("support/dashboard")
    suspend fun getSupportDashboard(): Response<SupportDashboardDto>

    @GET("payments/relay-cash/dashboard")
    suspend fun getRelayCashDashboard(): Response<RelayCashDashboardDto>

    @GET("support/tickets")
    suspend fun getSupportTickets(
        @Query("status") status: String? = null,
        @Query("search") search: String? = null,
        @Query("limit") limit: Int? = 50,
    ): Response<List<SupportTicketDto>>

    @GET("support/tickets/{id}")
    suspend fun getSupportTicket(@Path("id") id: String): Response<SupportTicketDetailDto>

    @POST("support/tickets/{id}/reply")
    suspend fun replySupportTicket(
        @Path("id") id: String,
        @Body request: SupportReplyRequest,
    ): Response<SuccessResponse>

    @POST("support/tickets/{id}/status")
    suspend fun updateSupportTicketStatus(
        @Path("id") id: String,
        @Body request: SupportStatusRequest,
    ): Response<SuccessResponse>
}
